st2actions/st2actions/scheduler/handler.py

```diff
@@ -60,25 +62,37 @@ def __init__(self):
         self._shutdown = False
         self._pool = eventlet.GreenPool(size=cfg.CONF.scheduler.pool_size)
         self._coordinator = coordination_service.get_coordinator()
+        self._main_thread = None
+        self._cleanup_thread = None
 
     def run(self):
-        LOG.debug('Entering scheduler loop')
+        LOG.debug('Starting scheduler handler...')
 
         while not self._shutdown:
             eventlet.greenthread.sleep(cfg.CONF.scheduler.sleep_interval)
+            self.process()
 
-            execution_queue_item_db = self._get_next_execution()
+    @retrying.retry(
+        retry_on_exception=service_utils.retry_on_exceptions,
+        stop_max_attempt_number=cfg.CONF.scheduler.retry_max_attempt,
+        wait_fixed=cfg.CONF.scheduler.retry_wait_msec)
+    def process(self):
+        execution_queue_item_db = self._get_next_execution()
 
-            if execution_queue_item_db:
-                self._pool.spawn(self._handle_execution, execution_queue_item_db)
+        if execution_queue_item_db:
+            self._pool.spawn(self._handle_execution, execution_queue_item_db)
 
     def cleanup(self):
-        LOG.debug('Starting scheduler garbage collection')
+        LOG.debug('Starting scheduler garbage collection...')
 
         while not self._shutdown:
             eventlet.greenthread.sleep(cfg.CONF.scheduler.gc_interval)
             self._handle_garbage_collection()
 
+    @retrying.retry(
```

Comment:
I'm personally still not too sure about this retry here. It's seems like a one off - aka we don't do it in other similar services.

Reply:
Besides that, the linking looks good to me, let's please just add some test cases for it.

Reply:
I think having the retries there is ok.  For a single server install w/o any complex service management, this prevents temporary hiccups with MongoDB connection.  If users want to fail fast, they can reconfigure the retries.  If you're worry about consistency, we can revisit.  After this issue, our service pattern need an overhaul.